docs/development.rst

```diff
@@ -323,3 +328,104 @@ Where ``desiredVersion`` is the version that will be written to the
+local branch has diverged from its remote tracking branch, running ``git
+pull`` will create a merge commit locally to join the two branches.
+
+In some workflows, this is not an issue.  For us, however, it creates a
+problem in the future.  When you are ready to submit your topic branch in a
+pull request, we ask you to squash your commits (usually down to one
+commit).  Given the complex graph topography created by all of the merges, the
+order in which git applies commits in the squash is very difficult to
+reason about and will likely create merge conflicts that you find
+unnecessary and nonsensical (and therefore, highly aggravating!).
+
+We instead recommend using ``git fetch`` and ``git rebase`` to update your
+local topic branch against a mainline branch.  This will create a linear
+commit history in your topic branch, which will be easy to squash, since the
+commits are applied in the squash in the order that you made them.
+
+``git pull`` does have the ``--rebase`` option which will do a rebase
+instead of a merge to incorporate the remote branch.  One can also set the
+``branch.autosetuprebase always`` config option to have ``git pull`` do a
+rebase by default (i.e. without passing the ``--rebase`` flag) instead of a
+merge.  This will avoid the issue of squashing a non-linear commit history.
+
+So, in truth, we are really recommending against squashing local branches
+with many merge commits in them.  However, using the default settings for
+``git pull`` is the easiest way to end up in this situation.  Therefore,
+don't use ``git pull`` unless you know what you are doing.
+
++++++++++++++++++++
+Squash, then rebase
++++++++++++++++++++
+
+When updating a local topic branch with changes from a mainline branch, we
+recommend squashing commits in your topic branch down to one commit before
+rebasing on top of the mainline branch.  The reason for this is that, under the
+hood, to apply the rebase ``git rebase`` essentially cherry-picks each
+commit from your topic branch not in the mainline branch and applies it to the
+mainline branch.  Each one of these applications can cause a merge
+conflict.  It is much better to face the potential of only one merge
+conflict than N merge conflicts (where N is the number of unique commits in the
+local branch)!
+
+The difficulty of proceeding the opposite way (rebasing, then squashing) is
+only compounded because of the unintuitiveness of the N merge conflicts.
+When presented with a merge conflict, your likely intuition is to put the
+file in the state that you think it ought to be in, namely the condition it was
+in after the Nth commit.  However, if that state was different than the
+state that git thinks it should be in -- namely, the state of the file at
+commit X where X<N -- then you have only created the potential for more
+merge conflicts.  When the next intermediate commit, Y (where X<Y<N) is
+applied, it too will create a merge conflict.  And so on.
+
```

Comment:
This is very good advice, which I didn't know about. However, I don't know how you'd actually do it -- can you put in a command example of the commands required to squash a topic branch, and then rebase?